File: 007.跨端APP应用（移动端开发工程师）/uniapp-project/utils/finance.ts
import type { AccountType } from '@/api/accounts'
import type { CategoryType } from '@/api/categories'

/**
 * V1.2 金额核算审计:ISO 4217 货币码 → 显示符号。
 * 没收录的码回退 `¥`,与历史行为一致。
 */
export function currencySymbol(code: string | null | undefined): string {
  switch ((code ?? '').toUpperCase()) {
    case 'CNY':
    case 'RMB': return '¥'
    case 'USD': return '$'
    case 'EUR': return '€'
    case 'GBP': return '£'
    case 'JPY': return '¥'
    case 'HKD': return 'HK$'
    default: return '¥'
  }
}

/**
 * V1.2 金额核算审计:
 *   - NaN/Infinity → '--' 而不是 '¥NaN' 炸屏。
 *   - currency 优先于 withSymbol;传 currency 时直接用货币码对应符号。
 */
export function formatAmount(
  n: number | null | undefined,
  withSymbol = false,
  currency?: string | null,
): string {
  if (n == null || !Number.isFinite(n)) return '--'
  const num = n as number
  const body = num.toLocaleString('zh-CN', { minimumFractionDigits: 2, maximumFractionDigits: 2 })
  const sym = currency != null ? currencySymbol(currency) : (withSymbol ? '¥' : '')
  return sym + body
}

/** ISO 字符串(后端返回的 createdAt / updatedAt)格式化为本地时区 `YYYY-MM-DD HH:mm`。 */
export function formatDateTime(iso: string): string {
  const d = new Date(iso)
  if (Number.isNaN(d.getTime())) return iso
  const pad = (n: number) => String(n).padStart(2, '0')
  return `${d.getFullYear()}-${pad(d.getMonth() + 1)}-${pad(d.getDate())} ${pad(d.getHours())}:${pad(d.getMinutes())}`
}

export function typeOfAccount(t: AccountType): string {
  return ({ cash: '现金', debit: '借记卡', credit: '信用卡', wallet: '钱包', investment: '投资', other: '其他' } as const)[t] ?? '其他'
}

export function typeOfCategory(t: CategoryType): 'expense' | 'income' | 'transfer' {
  return t
}

export function balanceSign(n: number): -1 | 0 | 1 {
  return n < 0 ? -1 : n > 0 ? 1 : 0
}

File: 007.跨端APP应用（移动端开发工程师）/flutter-project/lib/core/utils/finance.dart
import 'package:intl/intl.dart';

import '../api/models.dart';

/// V1.2 金额核算审计:ISO 4217 货币码 → 显示符号。
/// 没收录的码(空/null)回退 `¥`,与历史行为一致。
String currencySymbol(String? code) {
  switch (code?.toUpperCase()) {
    case 'CNY':
    case 'RMB':
      return '¥';
    case 'USD':
      return r'$';
    case 'EUR':
      return '€';
    case 'GBP':
      return '£';
    case 'JPY':
      return '¥';
    case 'HKD':
      return 'HK\$';
    default:
      return '¥';
  }
}

/// 对齐 utils/finance.ts。
///
/// V1.2 金额核算审计:
///   - 入参若 NaN / Infinity / null,返回 `'--'` 而不是渲染 `¥NaN`(避免炸屏)。
///   - `currency` 优先于 `withSymbol`;传 `currency` 时直接用货币码对应的符号,
///     否则按 `withSymbol` 兜底为 `¥`。
String formatAmount(num? n, {bool withSymbol = false, String? currency}) {
  if (n == null || !n.isFinite) return '--';
  final formatter = NumberFormat('#,##0.00', 'zh_CN');
  final sym = currency != null ? currencySymbol(currency) : (withSymbol ? '¥' : '');
  return sym + formatter.format(n);
}

/// ISO 字符串 -> `YYYY-MM-DD HH:mm`(本地时区)。
/// 与原 utils/formatDateTime 一致;parse 失败返回原字符串。
String formatDateTime(String iso) {
  try {
    final d = DateTime.parse(iso).toLocal();
    final pad = (int n) => n.toString().padLeft(2, '0');
    return '${d.year}-${pad(d.month + 1 - 1)}-${pad(d.day)} ${pad(d.hour)}:${pad(d.minute)}';
  } catch (_) {
    return iso;
  }
}

String typeOfAccount(AccountType t) {
  switch (t) {
    case AccountType.cash:
      return '现金';
    case AccountType.debit:
      return '借记卡';
    case AccountType.credit:
      return '信用卡';
    case AccountType.wallet:
      return '钱包';
    case AccountType.investment:
      return '投资';
    case AccountType.other:
      return '其他';
  }
}

/// recordType 字面量,与 uniapp 中 utils/finance.typeOfCategory 等价。
String typeOfCategory(CategoryType t) => t.name;

int balanceSign(num n) => n < 0 ? -1 : (n > 0 ? 1 : 0);
